h5py/_hl/compat.py

```diff
@@ -38,11 +37,10 @@ def filename_decode(filename):
     called on any filenames passed from the HDF5 library. See the documentation
     on filenames in h5py for more information.
     """
-    if sys.platform == "win32":
-        if isinstance(filename, bytes):
-            return filename.decode(WINDOWS_ENCODING, "strict")
-        elif isinstance(filename, str):
-            return filename
-        else:
-            raise TypeError("expect bytes or str, not %s" % type(filename).__name__)
-    return fsdecode(filename)
+    if not isinstance(filename, (str, bytes)):
+        raise TypeError(f"expect bytes or str, not {type(filename).__name__}")
```

Comment:
moving this `raise` statement up here isn't a API-breaking change: `os.fsdecode` would already have raised a `TypeError` on any platform in the same conditions, but this change makes the exact error message and traceback consistent on all platforms.